wgengine/magicsock/relaymanager.go

```diff
@@ -337,26 +346,68 @@ func (r *relayManager) stopWorkRunLoop(ep *endpoint, f stopHandshakeWorkFilter)
 			_, knownServer := r.serversByDisco[disco]
 			if knownServer || f == stopHandshakeWorkAllServers {
 				handshakeWork.cancel()
-				handshakeWork.wg.Wait()
-				delete(byServerDisco, disco)
-				delete(r.handshakeWorkByServerDiscoVNI, serverDiscoVNI{handshakeWork.se.ServerDisco, handshakeWork.se.VNI})
+				done := <-handshakeWork.doneCh
+				r.handleHandshakeWorkDoneRunLoop(done)
 			}
 		}
-		if len(byServerDisco) == 0 {
-			delete(r.handshakeWorkByEndpointByServerDisco, ep)
-		}
 	}
 }
 
-func (r *relayManager) handleRxChallengeRunLoop(challenge relayHandshakeChallengeEvent) {
-	work, ok := r.handshakeWorkByServerDiscoVNI[serverDiscoVNI{challenge.disco, challenge.vni}]
-	if !ok {
+// addrPortVNI represents a combined netip.AddrPort and Geneve header virtual
+// network identifier.
+type addrPortVNI struct {
+	addrPort netip.AddrPort
+	vni      uint32
+}
+
+func (r *relayManager) handleRxHandshakeDiscoMsgRunLoop(event relayHandshakeDiscoMsgEvent) {
+	var (
+		work *relayHandshakeWork
+		ok   bool
+	)
+	apv := addrPortVNI{event.from, event.vni}
+	switch event.msg.(type) {
+	case *disco.BindUDPRelayEndpointChallenge:
+		work, ok = r.handshakeWorkByServerDiscoVNI[serverDiscoVNI{event.disco, event.vni}]
+		if !ok {
+			// No outstanding work tied to this challenge, discard.
+			return
+		}
+		_, ok = r.handshakeWorkAwaitingPong[work]
+		if ok {
+			// We've seen a challenge for this relay endpoint previously,
+			// discard. Servers only respond to the first src ip:port they see
+			// binds from.
+			return
+		}
+		_, ok = r.addrPortVNIToHandshakeWork[apv]
+		if ok {
+			// There is existing work for the same [addrPortVNI] that is not
+			// 'work'. If both instances happen to be on the same server we
+			// could attempt to resolve event order using LamportID. For now
+			// just leave both work instances alone and take no action other
+			// than to discard this challenge msg.
+			return
+		}
+		// Update state so that future ping/pong will route to 'work'.
+		r.handshakeWorkAwaitingPong[work] = apv
+		r.addrPortVNIToHandshakeWork[apv] = work
+	case *disco.Ping, *disco.Pong:
+		work, ok = r.addrPortVNIToHandshakeWork[apv]
+		if !ok {
+			// No outstanding work tied to this [addrPortVNI], discard.
+			return
+		}
```

Comment:
Checking my understanding: case arm doesn't do anything with pings/pongs. I believe this is expected right now, and an upcoming change will plumb relay addresses into magicsock best-path selection, at this location among others. Is that accurate?

Reply:
The ping/pong case verifies there is associated work for the addrport and VNI, and we fall through to the select on L405 if so. That select can write to a channel that is read by `handshakeServerEndpoint()`, which is responsible for handshaking and probing (maybe this should be part of the naming) via ping/pong.

When `handshakeServerEndpoint()` returns the relay endpoint is either functional (we rx'd a pong from the remote end post-handshake with the relay) or not. We evaluate its outcome in `handleHandshakeWorkDoneRunLoop()` which is where the best path plumbing TODO you are referring to is at the moment:

> // This relay endpoint is functional.
// TODO(jwhited): Set it on done.work.ep.bestAddr if it is a betterAddr().